File: codes/perl/polymorphism/overloading.pl
package Person;
sub new { return bless {}, shift; }
sub read {
    if (@_< 2 ) {
        print "I am reading\n";
        return;
    }
    my ($self, $arg1) = @_;
    print "I am reading: $arg1\n";
}
1;

package Student;
our @ISA = qw(Person);
sub read {
    if (@_ == 3) {
        my ($self, $arg1, $arg2) = @_;
        print "I am reading: $arg1 for $arg2 times\n";
    }
    else {
        $self = shift;
        $self->SUPER::read(@_);
    }
}
1;

print "PERSON\n";
$pe = Person->new();
$pe->read();
$pe->read("Book");
$pe->read("Book", 5);

print "STUDENT\n";
$st = Student->new();
$st->read();
$st->read("Book");
$st->read("Book", 5);

File: codes/php/polymorphism/overloading.php
<?php

class Person {
    public function read(){
        switch (func_num_args()) {
            case 0:
                echo "I am a reading\n";
                break;
            case 1:
                $arg1 = func_get_arg(0);
                if (gettype($arg1) == "string"){
                    echo "A text: $arg1\n";
                }
                break;
            default:
        }
    }
}

class Student extends Person {
    public function read(...$args){
        if (count($args) == 1 && gettype($args[0]) == "integer"){
            echo "I read on table n°: ". $args[0] . "\n";
        }
        else { parent::read(...$args); }
    }
}

echo "PERSON\n";
$pe = new Person();
$pe->read();
$pe->read("I am a person");

echo "STUDENT\n";
$st = new Student();
$st->read();
$st->read("I am a student");
$st->read(5);

?>
